File: Application/Api/Orders/List/Handlers/ListOrdersHandler.cs
using Application.ErrorHandling.Application;
using Application.Interfaces.Persistence;
using Domain.ValueObjects.Order;
using MediatR;
using OneOf;

namespace Application.Api.Orders.List.Handlers;

public class ListOrdersHandler : IRequestHandler<ListOrdersQuery, OneOf<ListOrdersResult, List<PlainApplicationError>>>
{
    private readonly IOrderRepository _orderRepository;

    public ListOrdersHandler(IOrderRepository orderRepository)
    {
        _orderRepository = orderRepository;
    }

    public async Task<OneOf<ListOrdersResult, List<PlainApplicationError>>> Handle(ListOrdersQuery request, CancellationToken cancellationToken)
    {
        Tuple<string, bool>? orderBy = new Tuple<string, bool>("DateCreated", false);
        if (request.OrderBy == "newest")
        {
            orderBy = new Tuple<string, bool>("DateCreated", false);
        } 
        else if (request.OrderBy == "oldest")
        {
            orderBy = new Tuple<string, bool>("DateCreated", true);
        }
        else if (request.OrderBy == "total desc")
        {
            orderBy = new Tuple<string, bool>("Total", false);
        }
        else if (request.OrderBy == "total asc")
        {
            orderBy = new Tuple<string, bool>("Total", true);
        }

        var orders = await _orderRepository.FindAllAsync(
            minTotal: request.MinTotal,
            maxTotal: request.MaxTotal,
            status: (request.Status is null || !OrderStatus.IsValid(request.Status)) ? null : new OrderStatus(name: request.Status),
            createdBefore: request.CreatedBefore,
            createdAfter: request.CreatedAfter,
            id: request.Id,
            productId: request.ProductId,
            productHistoryId: request.ProductHistoryId,
            orderBy: orderBy
        );

        var result = new ListOrdersResult(orders: orders);
        return result;
    }
}

File: frontend/src/application/commands/orders/listOrders/ListOrdersHandler.ts
import { ICommandHandler } from "../../ICommandHandler";
import { err, ok } from "neverthrow";
import IListProductsCommand from "./ListOrdersCommand";
import IListProductsResult from "./IListOrdersResult";
import IOrderDataAccess from "../../../interfaces/dataAccess/IOrderDataAccess";
import IListOrdersResult from "./IListOrdersResult";
import ListOrdersCommand from "./ListOrdersCommand";
import OrderStatus from "../../../../domain/valueObjects/Order/OrderStatus";

export default class ListOrdersHandler implements ICommandHandler<IListProductsCommand, IListProductsResult> {
    private _orderDataAccess: IOrderDataAccess;

    constructor(props: { orderDataAccess: IOrderDataAccess }) {
        this._orderDataAccess = props.orderDataAccess;
    }

    async handle(request: ListOrdersCommand): Promise<IListOrdersResult> {
        try {
            if (request.status != null && OrderStatus.isValid(request.status)) {
                request.status = null;
            }

            const result = await this._orderDataAccess.listOrders({
                id: request.id,
                createdAfter: request.createdAfter,
                createdBefore: request.createdBefore,
                maxTotal: request.maxTotal,
                minTotal: request.minTotal,
                status: request.status,
                productId: request.productId,
            });

            if (result.isErr()) {
                return err({ type: "API", data: result.error });
            }

            return ok({ orders: result.value });
        } catch (error: unknown) {
            return err({ type: "Exception", data: error });
        }
    }
}
